src/Microsoft.AspNet.OData.Shared/ODataAPIHandler.cs

```diff
@@ -64,129 +69,245 @@ public abstract class ODataAPIHandler<TStructuralType>: IODataAPIHandler where T
-        /// </summary>
-        /// <param name="resource">resource to apply odata id on.</param>
+        /// <param name="resource">Resource to execute.</param>
         /// <param name="model">The model.</param>
-        internal virtual void UpdateLinkedObjects(TStructuralType resource, IEdmModel model)
+        /// <param name="apiHandlerFactory">API Handler Factory.</param>
+        public virtual void DeepInsert(TStructuralType resource, IEdmModel model, ODataAPIHandlerFactory apiHandlerFactory)
         {
             if (resource != null && model != null)
             {
-                this.ParentObject = resource;
-                CheckAndApplyODataId(resource, model);
+                CopyObjectProperties(resource, model, this, apiHandlerFactory);
             }
         }
 
-        private static ODataPath GetODataPath(string path, IEdmModel model)
+        internal static void CopyObjectProperties(object obj, IEdmModel model, IODataAPIHandler apiHandler, ODataAPIHandlerFactory apiHandlerFactory)
         {
-            ODataUriParser parser = new ODataUriParser(model, new Uri(path, UriKind.Relative));
-            ODataPath odataPath = parser.ParsePath();
+            Type type = obj.GetType();
+            PropertyInfo[] properties = type.GetProperties();
+            PropertyInfo odataIdContainerProperty = properties.FirstOrDefault(s => s.PropertyType == typeof(ODataIdContainer));
 
-            return odataPath;
-        }
+            string edmFullName = type.EdmFullName();
+            IEdmSchemaType schemaType = model.FindType(edmFullName);
+            IEdmEntityType edmEntityType = schemaType as IEdmEntityType;
 
-        private void CheckAndApplyODataId(object obj, IEdmModel model)
-        {
-            Type type = obj.GetType();
-            PropertyInfo property = type.GetProperties().FirstOrDefault(s => s.PropertyType == typeof(ODataIdContainer));
-            if (property != null && property.GetValue(obj) is ODataIdContainer container && container != null)
+            IEnumerable<IEdmStructuralProperty> entityKey = edmEntityType.Key();
+
+            IDictionary<string, object> keys = GetKeys(entityKey, obj, type);
+
+            IODataAPIHandler odataIdContainerHandler = null;
+
+            if (odataIdContainerProperty != null && odataIdContainerProperty.GetValue(obj) is ODataIdContainer container && container != null && apiHandlerFactory != null)
             {
                 ODataPath odataPath = GetODataPath(container.ODataId, model);
-                object res = ApplyODataIdOnContainer(obj, odataPath);
-                foreach (PropertyInfo prop in type.GetProperties())
+
+                if (odataPath != null)
                 {
-                    object resVal = prop.GetValue(res);
+                    odataIdContainerHandler = apiHandlerFactory.GetHandler(odataPath);
 
-                    if (resVal != null)
+                    if (odataIdContainerHandler != null)
```

Comment:
here too. what happens if `odataIdContainerHandler` is null.

Reply:
That entire section handles scenarios where we have an ODataIdContainer. 

If we don't have the ODataIdContainer, or we can't get the correct ODataPath or we can't get the handler from the path, we cannot call `TryGet` and `TryAddRelatedObject` in the `ApplyHandlers` method